[14.0] OCA' Odoo instance docker image (#59)
OCA/oca-custom: CONTRIBUTE.md

# Contribute to the OCA's Odoo instance

This guide aims to help happy volunteers to contribute to the OCA's Odoo instance. I
suppose mainly the OCA's Internal Tools team.

It's split into 3 sections:

- [Concepts](#concepts): Main concepts to understand and general organization
- [Processes](#processes): Helping doing the work without missing crucial steps
- [HowTos](#howto): How to do specific tasks

## Concepts

> This repository is setup as other OCA's repositories to launch CI as usual and as an
> extra configuration in order to realease the OCA' Docker image used by our Odoo
> instance.

Reviewer: ```suggestion
This repository is setup as other OCA's repositories to launch CI as usual and as an
extra configuration in order to build the OCA' Docker image used by our Odoo
instance, as well as facilitate the bootstrapping of a development environment.
```

Managing and freezing modules versions rely on python tools:

- [uv](https://docs.astral.sh/uv/)
- [hatch-odoo](https://pypi.org/project/hatch-odoo/)

## Processes

Here we focus on what to do without explaining how to do it.

## HowTos

Here we focus on how to do it, it's a suggest way to works but feel free to use your own
way.

### Setup developer environment

Requirements:

- Postrgesql
- [uv](https://docs.astral.sh/uv/)
- Some dependencies to be able to build some python packages: `libpq-dev`,

>   `build-essential`, TODO

Reviewer: + wkhtmltopdf

> Prepare a python virtual environment and install the required dependencies:

Reviewer: ```suggestion
Prepare a python virtual environment with the correct python version (which uv will download for you if necessary) and install the required dependencies:
```

```bash
uv sync
```

### Setup database and launch tests

- setup database with demo data and all oca modules installed

```bash
uv run odoo -d oca-custom -i oca_all --stop-after-init --without-de
```

- run tests using pytest launcher

````bash
uv run pytest --odoo-database oca-custom --cov ./oca_psc_team/ oca_psc_team/
``

### Update OCB Branch

```bash

> uvx --from git-aggregator gitaggregate -c repos.yaml -d ./src/ocb -p

Reviewer: ```suggestion
```

uv sync -P odoo
````

### Update a specific OCA module dependency using the latest pypi release

```bash

> uv sync -P odoo14-addon-<module-name-underscores-replaced-by-hyphens>

Reviewer: ```suggestion
uv sync --upgrade-package odoo14-addon-<module_name>
```

No need to replace underscores here?

Author: I'll give a try, I through it's the name of the package which is declared with hyphens in python project

## Concepts

This repository is setup as other OCA's repositories to launch CI as usual and as an
extra configuration in order to build the OCA' Docker image used by our Odoo instance,
as well as facilitate the bootstrapping of a development environment.

Managing and freezing modules versions rely on python tools:

- [uv](https://docs.astral.sh/uv/)
- [hatch-odoo](https://pypi.org/project/hatch-odoo/)

## Processes

Here we focus on what to do without explaining how to do it.

### Release

- Go to action tab and click on `Release OCA`
- click on `Run Workflow` dropdown menu
- click on `Run workflow`

This will:

- change version in pyproject on the head of branch 14.0
- add a git tag with the new version autogenerated
- push on branch 14.0 and push the new tag which will trigger other flow to create the
  docker image
- A new docker image should be available in the
  [github docker registry](https://github.com/oca/oca-custom/pkgs/container/oca-custom)

### deployment

Ask administrator to deploy the given version.

## HowTos

Here we focus on how to do it, it's a suggest way to works but feel free to use your own
way.

### Setup developer environment

Requirements:

> - Postrgesql

Reviewer 2: ```suggestion
- Postgresql
```

- [uv](https://docs.astral.sh/uv/)
- Some dependencies to be able to build some python packages: `libpq-dev`,
  `build-essential`, TODO
- wkhtmltopdf

Prepare a python virtual environment with the correct python version (which uv will
download for you if necessary) and install the required dependencies:

```bash
uv sync
```

> 

Reviewer 2: ```suggestion

### Development

For addons living in this repository, you can just change code and restart Odoo with the `uv run` command.

For addons in other repositories, the procedure is as follows:

- check out the repository somewhere, ie /src/$repo
- add the following line to ``pyproject.toml`` in the `[tool.uv.sources]` section:

        odoo14-addon-$youraddon = { path = "/srv/$repo/setup/$youraddon", editable = true }
     
- run `uv sync`
- restart Odoo
```

### Setup database and launch tests

- setup database with demo data and all oca modules installed

```bash
uv run odoo -d oca-custom -i oca_all --stop-after-init --without-de
```

- run tests using pytest launcher

````bash
uv run pytest --odoo-database oca-custom --cov ./oca_psc_team/ oca_psc_team/
``

### Update OCB Branch

```bash
uv sync -P odoo
````

### Update a specific OCA module dependency using the latest pypi release

```bash
uv sync -P odoo14-addon-<module-name>
```

### Bump all dependencies to the latest version

```bash
uv sync -U

> ```

Reviewer 2: I must admit I've never heard of `uv` before looking at this PR, so probably dumb question: What's the recommended way of editing code here?

Now I'm toying with this to get familiar with the process, I do

```sh
git clone $repo -b 14.0
cd $repo
pip install -e .
# do my stuff
# later
git commit -m $message && git push ...
```

is that's how it's meant to go?

Reviewer: Hi Holger, you probably missed the *Setup developer environment* section that this PR adds to the README.
Please have a look and if that is still unclear we should improve that section.

Reviewer 2: that's what I'm commenting on. It tells me to run `uv sync`, which I did. Now I've all the modules available in the project's python path, that's great, but what's the recommended way to actually do development now?

Reviewer: `uv sync` has created a virtualenv named `.venv` in the project directory.

The project itself is installed in editable mode and the addons are in addons_path, so you can simply edit the code, and run odoo with `uv run odoo` as mentioned above. This is more or less equivalent to running `.venv/bin/python -m odoo`, so you can configure your favorite IDE/editor/debugger with that.

If you add dependencies, just run `uv sync` again. `uv run odoo` will also re-sync automatically.

`uv` does not install `pip`, but you can run `uv pip list` to inspect what is installed in the venv.

Reviewer 2: thanks, so far it's clear. But I'm talking about developing stuff for addons in other repos, which will probably mostly happen - those are just in the python path to my understanding?

Let's say OCA asks for some patch to mis_builder, what do I do?

Reviewer: You add it to the [tool.uv.sources](https://docs.astral.sh/uv/reference/settings/#sources) section of `pyproject.toml` and `uv sync` / `uv run` again. There are a couple already with git references to branches and PRs, but I think you can add a local directory too.

So it would likely be something like

```toml
[tool.uv.sources]
odoo14-addon-mis_builder = { path = "/home/.../OCA/mis-builder/setup/mis_builder", editable = true }
```

which you then replace with a git reference when you have done the mis-builder PR.

Reviewer: And of course, odoo `--addons-path` continues to work normally too.

Reviewer 2: great, thanks. will play a bit with this and propose a section about that above
